core/asset/type.go

```diff
@@ -1,28 +1,79 @@
 package asset
 
+import (
+	"errors"
+	"regexp"
+)
+
+var (
+	errTypeInvalidLength    = errors.New("type length must be 3 to 16 inclusive")
+	errTypeInvalidCharacter = errors.New("type must be combination of alphanumeric and underscores")
+)
+
+var invalidTypePattern = regexp.MustCompile(`[^a-z0-9_]`)
+
 const (
-	TypeTable        Type = "table"
-	TypeJob          Type = "job"
-	TypeDashboard    Type = "dashboard"
-	TypeTopic        Type = "topic"
-	TypeFeatureTable Type = "feature_table"
-	TypeApplication  Type = "application"
-	TypeModel        Type = "model"
-	TypeQuery        Type = "query"
-	TypeMetric       Type = "metric"
+	typeMinLength = 3
+	typeMaxLength = 16
 )
 
-// AllSupportedTypes holds a list of all supported types struct
-var AllSupportedTypes = []Type{
-	TypeTable,
-	TypeJob,
-	TypeDashboard,
-	TypeTopic,
-	TypeFeatureTable,
-	TypeApplication,
-	TypeModel,
-	TypeQuery,
-	TypeMetric,
+const (
+	typeTable        Type = "table"
+	typeJob          Type = "job"
+	typeDashboard    Type = "dashboard"
+	typeTopic        Type = "topic"
+	typeFeatureTable Type = "feature_table"
+	typeApplication  Type = "application"
+	typeModel        Type = "model"
+	typeQuery        Type = "query"
+	typeMetric       Type = "metric"
+)
+
+var supportedTypes = []Type{
```

Comment:
is this still needed if we have below mapping? if this can be removed let's remove it

Reply:
actually we can use the map as you mentioned, but the only reason for not removing it is to keep the order. though, sure updating it to remove

Reply:
updated